libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp

```diff
@@ -0,0 +1,228 @@
+        }
+    }
+
+    read_temp();
+
+    read_serial_number();
+    read_remaining_capacity();
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_temp(void)
+{
+    uint16_t data;
+    if (read_word(BATTMONITOR_SMBUS_TEMP, data)) {
+        float temp = ((float)(data - 2731)) * 0.1f;
+        _state.temperature = temp;
+        _state.temperature_time = AP_HAL::millis();
+    }
+
+    return true;
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_remaining_capacity()
+{
+    if (_params._pack_capacity > 0) {
+        uint16_t data;
+        if (read_word(BATTMONITOR_SMBUS_REMAINING_CAPACITY, data)) {
+            float consumed = _params._pack_capacity - data;
+            if (consumed > 0 && consumed < _params._pack_capacity) {
+                _state.consumed_mah = consumed;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+
+// read_block - returns true if successful
+bool AP_BattMonitor_SMBus_SUI::read_block(uint8_t reg, uint8_t* data, uint8_t len) const
+{
+    // buffer to hold results (2 extra byte returned holding length and PEC)
+    uint8_t buff[len+2];
+
+    // read bytes
+    if (!_dev->read_registers(reg, buff, sizeof(buff))) {
+        return false;
+    }
+
+    // get length
+    uint8_t bufflen = buff[0];
+
+    // sanity check length returned by smbus
+    if (bufflen == 0 || bufflen > len) {
+        return false;
+    }
+
+    // check PEC
+    uint8_t pec = get_PEC(AP_BATTMONITOR_SMBUS_I2C_ADDR, reg, true, buff, bufflen+1);
+    if (pec != buff[bufflen+1]) {
+        return false;
+    }
+
+    // copy data (excluding PEC)
+    memcpy(data, &buff[1], bufflen);
+
+    // return success
+    return true;
+}
+
+// read_bare_block - returns true if successful
+bool AP_BattMonitor_SMBus_SUI::read_block_bare(uint8_t reg, uint8_t* data, uint8_t len) const
+{
+    // read bytes
+    if (!_dev->read_registers(reg, data, len)) {
+        return false;
+    }
+
+    // return success
+    return true;
+}
+
+void AP_BattMonitor_SMBus_SUI::read_cell_voltages()
+{
+    // read cell voltages
+    uint16_t voltbuff[SUI_MAX_CELL_READ];
+
+    if (!read_block(REG_CELL_VOLTAGE, (uint8_t *)voltbuff, sizeof(voltbuff))) {
+        return;
+    }
+    float pack_voltage_mv = 0.0f;
+        
+    for (uint8_t i = 0; i < MIN(SUI_MAX_CELL_READ, cell_count); i++) {
+        const uint16_t cell = voltbuff[i];
+        _state.cell_voltages.cells[i] = cell;
+        pack_voltage_mv += (float)cell;
+    }
+
+    if (cell_count >= SUI_MAX_CELL_READ) {
+        // we can't read voltage of all cells. get overall pack voltage to work out
+        // an average for remaining cells
+        uint16_t total_mv;
+        if (read_block(BATTMONITOR_SMBUS_VOLTAGE, (uint8_t *)&total_mv, sizeof(total_mv)) &&
+            total_mv >= pack_voltage_mv) {
+            for (uint8_t i = SUI_MAX_CELL_READ; i < cell_count; i++) {
+                _state.cell_voltages.cells[i] = (total_mv - pack_voltage_mv) / (cell_count - SUI_MAX_CELL_READ);
```

Comment:
Lifting this out to a const `uint16_t` before the for loop saves 12 bytes of flash, and should be faster.

Reply:
done